File: src/main/scala/cn/cjie/scala/akka/mapreduce/MasterActor.scala
package cn.cjie.scala.akka.mapreduce;


import akka.actor.Actor
import akka.actor.ActorRef
import akka.actor.Props
import akka.routing.RoundRobinRouter

/**
 * Created with IntelliJ IDEA.
 * User: hucj
 * Date: 14-10-27
 * Time: 下午5:11
 * To change this template use File | Settings | File Templates.
 */
class MasterActor extends Actor {
    val mapActor = context.actorOf(Props[MapActor].withRouter(
            RoundRobinRouter(nrOfInstances = 5)), name = "map")
    val reduceActor = context.actorOf(Props[ReduceActor].withRouter(
            RoundRobinRouter(nrOfInstances = 5)), name = "reduce")
    val aggregateActor = context.actorOf(Props[AggregateActor],
            name = "aggregate")
    def receive: Receive = {
        case line: String =>
            mapActor ! line
        case mapData: MapData =>
            reduceActor ! mapData
        case reduceData: ReduceData =>
            aggregateActor ! reduceData
        case Result =>
            aggregateActor forward Result
    }
}

File: src/main/java/com/cjie/akka/mapreduce/MasterActor.java
package com.cjie.akka.mapreduce;

import akka.actor.ActorRef;
import akka.actor.Props;
import akka.actor.UntypedActor;
import akka.routing.RoundRobinRouter;

/**
 * Created with IntelliJ IDEA.
 * User: hucj
 * Date: 14-10-27
 * Time: 下午3:50
 * To change this template use File | Settings | File Templates.
 */
public class MasterActor extends UntypedActor {
    ActorRef mapActor = getContext().actorOf(
            new Props(MapActor.class).withRouter(new
                    RoundRobinRouter(5)), "map");
    ActorRef reduceActor = getContext().actorOf(
            new Props(ReduceActor.class).withRouter(new
                    RoundRobinRouter(5)),"reduce");
    ActorRef aggregateActor = getContext().actorOf(
            new Props(AggregateActor.class), "aggregate");
    @Override
    public void onReceive(Object message) throws Exception {
        if (message instanceof String) {
            mapActor.tell(message,getSelf());
        } else if (message instanceof MapData) {
            reduceActor.tell(message,getSelf());
        } else if (message instanceof ReduceData) {
            aggregateActor.tell(message, getSelf());
        } else if (message instanceof Result) {
            aggregateActor.forward(message, getContext());
        } else
            unhandled(message);
    }
}
